### PR TITLE
Added possibility to allow trusted vendors to change code outside their namespace (#3156)

### LINKED ISSUE #3156: [feature]: Add the possibility to allow certain packages to change code outside their namespace
**Is your feature request related to a problem? Please describe.**
As an agency we sometimes need to change `@magento/package` code from within our own packages `@experius/package` as an example we need to change the way a part of pwa studio works or do a bugfix that's not yet in the current released version of PWA Studio.

A more concrete example would be [this](https://github.com/magento/pwa-studio/pull/3078) bugfix to the core. It's merged but wasn't shipped with 10.0. So I'd like to patch this file with a targetable for our projects from within our own package `@experius/ui`

Unfortunately this isn't allowed, we either need to have the same namspace `@magento/peregrine` or be inside the local project. This safety check is added for security reasons. But it's also limiting to agencies because we don't want to maintain default (meaning we need this for all our projects) targetables on 20 different projects inside the customers source.

**Describe the solution you'd like**
I'd love to see a way where the developer can explicitly tell webpack to also allow package x,y and z to change code from other packages. This way we are much more flexible but still keep the security as tight since the developer needs to manually allow a certain package in their local project's webpack.

**Describe alternatives you've considered**
-

**Additional context**
-

**Please let us know what packages this feature is in regards to:**
- [x] `venia-concept`
- [ ] `venia-ui`
- [x] `pwa-buildpack`
- [ ] `peregrine`
- [ ] `pwa-devdocs`
- [ ] `upward-js`
- [ ] `upward-spec`
- [ ] `create-pwa`

### PR DESCRIPTION
**Pull request by 0m3r**
Gives the possibility to add trusted vendors in package.json
```
"pwa-studio": {
    "targets": {
      "intercept": "./local-intercept.js"
    },
   "trusted-vendors":  [
      "@vendor"
   ]
```
and add custom checking into the file: packages/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.js _assertAllowedToTransform

## Description

Added possibility to allow trusted vendors to change code outside their namespace

## Related Issue

Closes #3156.

## Acceptance

@sirugh
@zetlen

### Verification Stakeholders

@jerschipper
@Jordaneisenburger

### Verification Steps

in your project root package.json 

```json
"pwa-studio": {
    "targets": {
      "intercept": "./local-intercept.js"
    },
   "trusted-vendors":  [
      "@vendor"
   ]
```

in your extension intercept.js ('@vendor/extension/lib/intercept.js')
```js
const { Targetables } = require('@magento/pwa-buildpack');
module.exports = targets => {
   const targetables = Targetables.using(targets);

   const HeaderComponent = targetables.reactComponent(
        '@magento/venia-ui/lib/components/Header/header.js'
    );
    
    HeaderComponent.insertAfterJSX(    
        '<SearchBar isOpen={isSearchOpen} ref={searchRef} />',
        `<span>INSERTED</span>`
    );
}
```

### CODE REVIEW

**Comment by fooman** [comment]
@0m3r I like the direction where this is going. I am a little bit afraid that having this part of the .env makes a bit harder than necessary to administer though. Ideally the decision to add a vendor as trusted is done on a per project basis rather than on a per environment basis. It could lead to hard to track down issues once you end up with a list longer than a few vendors.

Ideally this configuration could live in a file that is part of the checked in code for the project. packages.json might a good fit as part of this:

```
  "pwa-studio": {
    "targets": {
      "intercept": "./local-intercept.js"
    }
```

to

```
  "pwa-studio": {
    "targets": {
      "intercept": "./local-intercept.js"
    },
   "trusted_vendors":  [
      "@vendor"
   ]
```

**Comment by larsroettig** [comment]
> @0m3r I like the direction where this is going. I am a little bit afraid that having this part of the .env makes a bit harder than necessary to administer though. Ideally the decision to add a vendor as trusted is done on a per project basis rather than on a per environment basis. It could lead to hard to track down issues once you end up with a list longer than a few vendors.
> 
> Ideally this configuration could live in a file that is part of the checked in code for the project. packages.json might a good fit as part of this:
> 
> ```
>   "pwa-studio": {
>     "targets": {
>       "intercept": "./local-intercept.js"
>     }
> ```
> 
> to
> 
> ```
>   "pwa-studio": {
>     "targets": {
>       "intercept": "./local-intercept.js"
>     },
>    "trusted_vendors":  [
>       "@vendor"
>    ]
> ```

Hi @0m3r. I totally agree with @fooman from my point of view this would be very nice if we can do this via `package.json`. 
Currently, I will ask the core team about their opinion. 


kind regards, 

Lars

**Comment by 0m3r** [comment]
> Currently, I will ask the core team about their opinion.
what about core team opinion?

**Comment by larsroettig** [comment]
@supernova-at agreed with me on this yesterday would be cool if you can do it

**Comment by 0m3r** [comment]
@larsroettig review this commit https://github.com/magento/pwa-studio/pull/3266/commits/e29c51e11ac9167f1102a0d8b95ddac5e5816f94, please

**Review by larsroettig — CHANGES_REQUESTED** [review]
Hi @0m3r, 
The code looks good pls. Add **test coverage**, fix the two failing test cases and run *yarn run prettier*. 

```
packages/peregrine/lib/targets/__tests__/peregrine-targets.spec.js
packages/venia-ui/lib/targets/__tests__/venia-ui-targets.spec.js
```

Thank you in advance. Also, I recommend joining as Slack then we answer questions mutch faster

**https://devdocs.magento.com/community/resources.html**

**Comment by 0m3r** [comment]
Can I fix test fails on "my" code?

something like below
```js
        const fs = require('fs');
...
        const context = path.resolve(__dirname, '../../../../../');
        const configPath = path.resolve(context, 'package.json');
        try {
            if (fs.existsSync(configPath)) {
                const config = require(configPath)['pwa-studio'];
                return config && config['trusted_vendors'] ? config['trusted_vendors'] : [];
                //file exists
            }
        } catch(err) {
          // console.error(err)
        }
        return [];
```

**Comment by larsroettig** [comment]
Hi @jimbo. 
what do think about this PR like the idea and is very helpful for partners. 

kind regards.,

Lars

**Comment by larsroettig** [comment]
One thinks about Extension `package.json` if some extension it allows itself via 
```
trusted_vendors":  [
      "@vendor"
   ]
```
Should be prevented in every case a `yarn install` should be *not* enough to get access to this low level to react extension point for security reasons.   

**Verification Steps**
* Checks for `trusted_vendors"` will only be parsed from the root  project `package.json`

**Comment by jimbo** [comment]
> Hi @jimbo.
> what do think about this PR like the idea and is very helpful for partners.

I think it's great. We can certainly do more to expose components + API, as we do for talons, but this is a great way to relieve people's pain while we figure out those details.

File: packages/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.js [hunk 1]
```diff
@@ -167,6 +169,22 @@ class ModuleTransformConfig {
...
+        return requestorVendor.length > 0 && vendors.includes(requestorVendor);
+    }
+    _getTrustedExtensionVendors() {
+        const context = path.resolve(__dirname, '../../../../../');
+        const configPath = path.resolve(context, 'package.json');
+        if (!fs.existsSync(configPath)) {
+            return [];
+        }
+        const config = require(configPath)['pwa-studio'];
+        return config && config['trusted_vendors']
```
[the remark this answers is not in the document]
Comment (0m3r) [reply packages/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.js:184]: Fair enough

File: packages/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.js [hunk 2]
```diff
@@ -167,6 +169,22 @@ class ModuleTransformConfig {
     _isLocal(requestor) {
         return requestor === this._localProjectName;
     }
+    _isTrustedExtensionVendor(requestor) {
+        const vendors = this._getTrustedExtensionVendors();
+        const requestorVendor = requestor.split('/')[0];
+        return requestorVendor.length > 0 && vendors.includes(requestorVendor);
+    }
+    _getTrustedExtensionVendors() {
+        const context = path.resolve(__dirname, '../../../../../');
```
[the remark this answers is not in the document]
Comment (0m3r) [reply packages/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.js:178]: to be honest, I found something similar in the code and used it.
process.cwd() like me more.

[the remark this answers is not in the document]
Comment (0m3r) [reply packages/pwa-buildpack/lib/WebpackTools/ModuleTransformConfig.js]: https://github.com/magento/pwa-studio/pull/3266/commits/691cfdce9ccfd4b5399c55d3776bddf50fc7763d

**Comment by 0m3r** [comment]
> @0m3r Add support for monorepo and I think @larsroettig was also requesting test coverage if possible. Otherwise this looks perfect, nice work

about tests

I did not find ModuleTransformConfig.spec.js
Unfortunately, I don't have enough experience to write this from scratch.

**Comment by tjwiebell** [comment]
@0m3r Existing code appears to be tested by integration tests in other packages, there is no unit test. Instead of building all that out in this scope, I've created a manual test that we'll run during regression and aim to automate if it becomes a priority. Thank you for the contribution.

✅  QA Passed

**Review by tjwiebell — APPROVED** [review]

**Pull request merged**